mne_bids/report.py

```diff
@@ -0,0 +1,461 @@
+"""Make BIDS report from dataset and sidecar files."""
+# Authors: Adam Li <adam2392@gmail.com>
+#
+# License: BSD (3-clause)
+import json
+import os.path as op
+from pathlib import Path
+
+import numpy as np
+from mne.externals.tempita import Template, sub
+
+from mne_bids.config import DOI, ALLOWED_KINDS
+from mne_bids.datasets import fetch_brainvision_testing_data
+from mne_bids.tsv_handler import _from_tsv
+from mne_bids.utils import (make_bids_basename, get_kinds,
+                            get_entity_vals, _parse_ext,
+                            _find_matching_sidecar, _parse_bids_filename,
+                            BIDSPath)
+
+# functions to be used inside Template strings
+FUNCTION_TEMPLATE = """{{py:
+def _pretty_str(listed):
+    if not isinstance(listed, list):
+        listed = list(listed)
+
+    if len(listed) == 1:
+        return ','.join(listed)
+
+    return '{}, and {}'.format(', '.join(listed[:-1]), listed[-1])
+}}"""
+
+BIDS_DATASET_TEMPLATE = """{{if name == 'n/a'}}This{{else}}The{{name}}{{endif}}
+dataset was created with BIDS version {{bids_version}}
+by {{_pretty_str(authors)}}. This report was generated with
+MNE-BIDS ({{mne_bids_doi}}). """
+
+PARTICIPANTS_TEMPLATE = """There are {{n_subjects}} subjects amongst whom there are
+{{n_males}} males and {{n_females}} females ({{n_sex_unknown}} unknown).
+There are {{n_rhand}} right hand, {{n_lhand}} left hand
+and {{n_ambidex}} ambidextrous subjects.
+Their ages are {{min_age}}-{{max_age}} ({{mean_age}} +/- {{std_age}}
+with {{n_age_unknown}} unknown). """
+
+MODALITY_AGNOSTIC_TEMPLATE = """
+Data was acquired using a {{_pretty_str(system)}} system
+({{_pretty_str(manufacturer)}} manufacturer with line noise at
+{{_pretty_str(powerlinefreq)}} Hz{{if software_filters != 'n/a'}} using filters
+({{software_filters}}).{{else}}.{{endif}}
+Each dataset is {{min_record_length}} to {{max_record_length}} seconds,
+for a total of {{py:round(total_record_length, 2)}} seconds of data recorded
+({{mean_record_length}} +/- {{std_record_length}}).
+The dataset consists of {{n_sessions}} recording sessions
+({{_pretty_str(sessions)}}),
+{{n_scans}} total scans, {{n_chs}} channels ({{n_good}} are used and
+{{n_bad}} are removed from analysis). """
+
+IEEG_TEMPLATE = """There are {{n_ecog_chs}} ECoG and
+{{n_seeg_chs}} SEEG channels. """
```

Comment:
for MEG too, it is useful to say the gradiometers and magnetometers separately.

Reply:
I got rid of this for now since the modality-agnostic summaries were already quite challenging.